File: apps/coord/src/cf-access.ts
// Verifies Cloudflare Access assertions at the coordinator front door.
// The fetch handler and pairing RPC share one gate built from CoordConfig.
// JWKS material is cached here; malformed or unverifiable assertions fail closed.

import { signal } from "@roost/shared/diag";
import type { CoordConfig } from "@roost/shared/config";
import { b64urlDecode } from "@roost/shared/jwt-base";
import { log } from "@roost/shared/log";

export interface CloudflareAccessIdentity {
  readonly email: string;
  readonly subject: string;
}

export interface CloudflareAccessGate {
  /** Verified identity, or null when the assertion is absent or invalid. */
  verify(headers: Headers): Promise<CloudflareAccessIdentity | null>;
}

export const ACCESS_CLOCK_SKEW_MS = 60_000;
export const JWKS_TTL_MS = 15 * 60_000;
export const JWKS_REFETCH_MIN_INTERVAL_MS = 60_000;
export const MAX_ACCESS_EMAIL_UTF8_BYTES = 320;

const BASE64URL_RE = /^[A-Za-z0-9_-]+$/;
type RejectionReason =
  | "absent"
  | "malformed"
  | "unknown_kid"
  | "bad_signature"
  | "bad_issuer"
  | "bad_audience"
  | "expired"
  | "bad_claims";

interface AccessJwtHeader {
  readonly alg?: unknown;
  readonly kid?: unknown;
}

interface ParsedAssertion {
  readonly headerPart: string;
  readonly payloadPart: string;
  readonly signature: Uint8Array;
  readonly payload: Uint8Array;
  readonly kid: string;
}

interface JwksCache {
  readonly fetchedAtMs: number;
  readonly keys: Map<string, JsonWebKey>;
}

/** Build the verifier, or leave every surface unchanged when Access is off. */
export function createCloudflareAccessGate(cfg: CoordConfig): CloudflareAccessGate | null {
  const teamDomain = cfg.cfAccessTeamDomain;
  const audience = cfg.cfAccessAud;
  if (!teamDomain || !audience) return null;

  const issuer = `https://${teamDomain}`;
  const jwksUrl = `${issuer}/cdn-cgi/access/certs`;
  let jwksCache: JwksCache | null = null;
  let lastUnknownKidRefetchAt = Number.NEGATIVE_INFINITY;
  let jwksFetchInFlight: Promise<boolean> | null = null;

  const fetchJwks = async (): Promise<boolean> => {
    if (jwksFetchInFlight) return jwksFetchInFlight;

    const fetchPromise = Promise.resolve()
      .then(() => fetch(jwksUrl))
      .then(async (response) => {
        if (!response.ok) {
          throw new Error(`JWKS fetch returned HTTP ${response.status}`);
        }
        const body = await response.json() as unknown;
        const keys = parseJwks(body);
        if (keys === null) throw new Error("JWKS response did not contain a keys array");
        jwksCache = { fetchedAtMs: Date.now(), keys };
        return true;
      })
      .catch((error: unknown) => {
        log.warn("cf-access", "jwks_fetch_failed", { error });
        return false;
      })
      .finally(() => {
        jwksFetchInFlight = null;
      });
    jwksFetchInFlight = fetchPromise;
    return fetchPromise;
  };

  const findJwk = async (kid: string): Promise<JsonWebKey | null> => {
    const now = Date.now();
    let refreshed = false;
    if (jwksCache === null || now - jwksCache.fetchedAtMs >= JWKS_TTL_MS) {
      if (!await fetchJwks()) return null;
      refreshed = true;
    }

    let key = jwksCache?.keys.get(kid) ?? null;
    if (key !== null || refreshed) return key;

    if (now - lastUnknownKidRefetchAt < JWKS_REFETCH_MIN_INTERVAL_MS) return null;
    lastUnknownKidRefetchAt = now;
    if (!await fetchJwks()) return null;
    key = jwksCache?.keys.get(kid) ?? null;
    return key;
  };

  const verify = async (headers: Headers): Promise<CloudflareAccessIdentity | null> => {
    const assertion = headers.get("cf-access-jwt-assertion");
    if (assertion === null) return rejectAccess("absent");

    const parsed = parseAssertion(assertion);
    if (parsed === null) return rejectAccess("malformed");

    let jwk: JsonWebKey | null;
    try {
      jwk = await findJwk(parsed.kid);
    } catch {
      return rejectAccess("unknown_kid");
    }
    if (jwk === null) return rejectAccess("unknown_kid");

    let key: CryptoKey;
    try {
      key = await crypto.subtle.importKey(
        "jwk",
        jwk,
        { name: "RSASSA-PKCS1-v1_5", hash: "SHA-256" },
        false,
        ["verify"],
      );
    } catch {
      return rejectAccess("bad_signature");
    }

    const message = new TextEncoder().encode(`${parsed.headerPart}.${parsed.payloadPart}`);
    let signatureValid = false;
    try {
      signatureValid = await crypto.subtle.verify(
        { name: "RSASSA-PKCS1-v1_5" },
        key,
        asArrayBuffer(parsed.signature),
        asArrayBuffer(message),
      );
    } catch {
      return rejectAccess("bad_signature");
    }
    if (!signatureValid) return rejectAccess("bad_signature");

    const claims = parseJsonObject(parsed.payload);
    if (claims === null) return rejectAccess("bad_claims");
    const claimError = validateClaims(claims, issuer, audience);
    if (claimError !== null) return rejectAccess(claimError);

    return {
      email: claims.email as string,
      subject: claims.sub as string,
    };
  };

  return { verify };
}

function rejectAccess(reason: RejectionReason): null {
  signal("cf-access.rejected", { reason, cooldownKey: reason });
  return null;
}

function parseAssertion(assertion: string): ParsedAssertion | null {
  const parts = assertion.split(".");
  if (parts.length !== 3 || parts.some((part) => part.length === 0)) return null;
  const [headerPart, payloadPart, signaturePart] = parts as [string, string, string];
  const headerBytes = decodeBase64Url(headerPart);
  const payload = decodeBase64Url(payloadPart);
  const signature = decodeBase64Url(signaturePart);
  if (headerBytes === null || payload === null || signature === null) return null;

  const header = parseJsonObject(headerBytes) as AccessJwtHeader | null;
  if (header === null || header.alg !== "RS256" || typeof header.kid !== "string" || header.kid.length === 0) {
    return null;
  }
  return { headerPart, payloadPart, signature, payload, kid: header.kid };
}

function parseJwks(body: unknown): Map<string, JsonWebKey> | null {
  if (body === null || typeof body !== "object" || !("keys" in body)) return null;
  const keys = body.keys;
  if (!Array.isArray(keys)) return null;

  const parsed = new Map<string, JsonWebKey>();
  for (const candidate of keys) {
    if (candidate === null || typeof candidate !== "object" || !("kid" in candidate)) continue;
    const kid = candidate.kid;
    if (typeof kid !== "string" || kid.length === 0) continue;
    parsed.set(kid, candidate as JsonWebKey);
  }
  return parsed;
}

function decodeBase64Url(segment: string): Uint8Array | null {
  if (!BASE64URL_RE.test(segment) || segment.length % 4 === 1) return null;
  try {
    return b64urlDecode(segment);
  } catch {
    return null;
  }
}

function parseJsonObject(bytes: Uint8Array): Record<string, unknown> | null {
  try {
    const text = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
    const value: unknown = JSON.parse(text);
    if (value === null || typeof value !== "object" || Array.isArray(value)) return null;
    return value as Record<string, unknown>;
  } catch {
    return null;
  }
}

function validateClaims(
  claims: Record<string, unknown>,
  issuer: string,
  audience: string,
): RejectionReason | null {
  if (claims.iss !== issuer) return "bad_issuer";
  if (!Array.isArray(claims.aud) || !claims.aud.includes(audience)) return "bad_audience";

  const nowMs = Date.now();
  if (!isFiniteNumeric(claims.exp)) return "bad_claims";
  const expMs = (claims.exp as number) * 1000;
  if (!Number.isFinite(expMs)) return "bad_claims";
  if (expMs <= nowMs - ACCESS_CLOCK_SKEW_MS) return "expired";

  if (!isFiniteNumeric(claims.iat)) return "bad_claims";
  const issuedAtMs = (claims.iat as number) * 1000;
  if (!Number.isFinite(issuedAtMs) || issuedAtMs > nowMs + ACCESS_CLOCK_SKEW_MS) return "bad_claims";

  const notBefore = claims.nbf;
  if (notBefore !== undefined) {
    if (!isFiniteNumeric(notBefore)) return "bad_claims";
    const notBeforeMs = (notBefore as number) * 1000;
    if (!Number.isFinite(notBeforeMs) || notBeforeMs > nowMs + ACCESS_CLOCK_SKEW_MS) return "bad_claims";
  }

  if (
    typeof claims.email !== "string" ||
    claims.email.length === 0 ||
    new TextEncoder().encode(claims.email).byteLength > MAX_ACCESS_EMAIL_UTF8_BYTES
  ) {
    return "bad_claims";
  }
  if (typeof claims.sub !== "string" || claims.sub.length === 0) return "bad_claims";
  return null;
}

function isFiniteNumeric(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function asArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength) as ArrayBuffer;
}

File: crates/roost-coord/src/auth/cf_access.rs
//! Cloudflare Access: the one place a fronted coordinator and a direct one meet.
//!
//! Ported from `apps/coord/src/auth/cf-access.ts`. One question, one request:
//! does it carry a verified edge identity? The answer is an IDENTITY, never a
//! principal -- see [`CloudflareAccessIdentity`]. `install_cloudflare_jwks` is
//! the one wiring line `serve` owes this file.

use std::sync::{Arc, OnceLock};

use axum::http::HeaderMap;
use roost_host::{CoordConfig, b64url_decode};
use serde_json::Value;

use crate::auth::jwt_verify::VerifyClock;

pub use crate::auth::cf_access_keyring::RsaJwks;

/// The header a Cloudflare Access assertion arrives in.
pub const ACCESS_ASSERTION_HEADER: &str = "cf-access-jwt-assertion";

/// The provider string persisted beside a pairing request's edge identity.
pub const ACCESS_PROVIDER: &str = "cloudflare-access";

/// How much clock difference a claim's instant may carry (`cf-access.ts:20`).
/// Kept at v2's 60 s: Access mints at its own edge.
pub const ACCESS_CLOCK_SKEW_MS: i64 = 60_000;

pub const JWKS_TTL_MS: i64 = 15 * 60_000;

/// The floor between two refetches caused by an absent `kid`: without it,
/// anyone who knows a `kid` is missing picks this coordinator's request rate.
pub const JWKS_REFETCH_MIN_INTERVAL_MS: i64 = 60_000;

pub const MAX_ACCESS_EMAIL_UTF8_BYTES: usize = 320;

/// Why a request carries no usable front-door identity. These are the reasons,
/// not a rendering of them: each is a different fact, and only `BadSignature`
/// is a peer's doing.
#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
pub enum AccessRejection {
    #[error("no cloudflare access assertion")]
    Absent,
    /// Not three base64url segments, not JSON, `alg` not `RS256`, or no `kid`.
    #[error("malformed cloudflare access assertion")]
    Malformed,
    #[error("unknown cloudflare access key id")]
    UnknownKid,
    #[error("cloudflare access key set unreachable")]
    JwksUnreachable,
    /// Cloudflare's key did not sign these bytes.
    #[error("bad cloudflare access signature")]
    BadSignature,
    #[error("bad cloudflare access issuer")]
    BadIssuer,
    #[error("bad cloudflare access audience")]
    BadAudience,
    #[error("expired cloudflare access assertion")]
    Expired,
    /// A claim is missing, untyped, out of bounds, or not reportable text.
    #[error("bad cloudflare access claims")]
    BadClaims,
}

/// A front-door identity Cloudflare Access has signed for this coordinator.
///
/// THE ANSWER IS NEVER A PRINCIPAL: no `From<..> for Principal`, no
/// `principal()` accessor, no public field, so no caller can turn an edge
/// assertion into authority. An edge identity is provenance only -- pairing
/// records the email beside a request (`handlers-pairing.ts:92-94`) and
/// `DevicesList` shows it to every browser -- so a forged header changes what
/// an operator reads and nothing else. Signature is verified BEFORE claims.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CloudflareAccessIdentity {
    email: String,
    subject: String,
}

impl CloudflareAccessIdentity {
    /// The address Cloudflare reported, bounded by [`is_reportable_email`].
    #[must_use]
    pub fn email(&self) -> &str {
        &self.email
    }

    /// Cloudflare's stable identifier for the signed-in user.
    #[must_use]
    pub fn subject(&self) -> &str {
        &self.subject
    }

    /// The provider name persisted beside the identity.
    #[must_use]
    pub fn provider(&self) -> &'static str {
        ACCESS_PROVIDER
    }
}

/// Which keys sign Access assertions, and whether one of them signed these
/// bytes. One trait, not two: both halves are the same dependency pair, and it is
/// the seam that makes every claim decision below testable with no network.
#[async_trait::async_trait]
pub trait CloudflareJwks: Send + Sync {
    /// The JWK a `kid` names, or `None` when the key set does not hold it.
    async fn jwk(&self, issuer: &str, kid: &str) -> Result<Option<String>, String>;

    /// Whether the JWK spelled by `jwk` signs `signing_input` under RS256.
    fn verify_rs256(&self, jwk: &str, signing_input: &str, signature: &[u8]) -> bool;
}

/// The production key ring: Cloudflare's key set over HTTPS, cached per issuer,
/// and RSA-SHA256 verification against it. The cache lives here because the key
/// set is the network's: a gate that rebuilt it per request would be one
/// Cloudflare slowdown from refusing every pairing.

fn configured(config: &CoordConfig) -> Option<(&str, &str)> {
    let domain = config.cf_access_team_domain.as_deref()?;
    let audience = config.cf_access_aud.as_deref()?;
    (!domain.trim().is_empty() && !audience.trim().is_empty()).then_some((domain, audience))
}

#[must_use]
pub fn cloudflare_access_configured(config: &CoordConfig) -> bool {
    configured(config).is_some()
}

/// Install the process's Access key ring, once, at boot. Until it is called a
/// coordinator with Access configured refuses every assertion rather than
/// believe one.
pub fn install_cloudflare_jwks(jwks: Arc<dyn CloudflareJwks>) -> Result<(), ()> {
    KEY_RING.set(jwks).map_err(|_| ())
}

static KEY_RING: OnceLock<Arc<dyn CloudflareJwks>> = OnceLock::new();

/// The verified edge identity of one request, or why there is not one.
pub async fn verify_edge_identity(
    config: &CoordConfig,
    headers: &HeaderMap,
    clock: VerifyClock,
) -> Result<Option<CloudflareAccessIdentity>, AccessRejection> {
    let Some((domain, audience)) = configured(config) else {
        return Ok(None);
    };
    let ring = KEY_RING.get().ok_or(AccessRejection::JwksUnreachable)?;
    let assertion = headers
        .get(ACCESS_ASSERTION_HEADER)
        .and_then(|value| value.to_str().ok())
        .ok_or(AccessRejection::Absent)?;
    let parsed = parse_assertion(assertion)?;
    let issuer = format!("https://{domain}");
    let jwk = ring
        .jwk(&issuer, &parsed.kid)
        .await
        .map_err(|_| AccessRejection::JwksUnreachable)?
        .ok_or(AccessRejection::UnknownKid)?;
    if !ring.verify_rs256(&jwk, &parsed.signing_input, &parsed.signature) {
        return Err(AccessRejection::BadSignature);
    }
    let (email, subject) = validate_claims(&parsed.payload, domain, audience, clock.now_ms)?;
    tracing::debug!(%subject, "cf_access.verified");
    Ok(Some(CloudflareAccessIdentity { email, subject }))
}

struct ParsedAssertion {
    signing_input: String,
    signature: Vec<u8>,
    payload: Vec<u8>,
    kid: String,
}

fn parse_assertion(assertion: &str) -> Result<ParsedAssertion, AccessRejection> {
    let parts: Vec<&str> = assertion.split('.').collect();
    let [header_part, payload_part, signature_part] = parts.as_slice() else {
        return Err(AccessRejection::Malformed);
    };
    if header_part.is_empty() || payload_part.is_empty() || signature_part.is_empty() {
        return Err(AccessRejection::Malformed);
    }
    let segment = |value: &str| b64url_decode(value).map_err(|_| AccessRejection::Malformed);
    let (header, payload, signature) = (
        segment(header_part)?,
        segment(payload_part)?,
        segment(signature_part)?,
    );
    // Exactly `RS256`: `none` and symmetric algorithms are refused before a key
    // is looked up.
    let header: Value = serde_json::from_slice(&header).map_err(|_| AccessRejection::Malformed)?;
    let alg = header
        .get("alg")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let kid = header
        .get("kid")
        .and_then(Value::as_str)
        .unwrap_or_default();
    if alg != "RS256" || kid.is_empty() {
        return Err(AccessRejection::Malformed);
    }
    Ok(ParsedAssertion {
        // The request's own bytes. Re-encoding from parsed values is how a
        // verifier checks a document nobody signed.
        signing_input: format!("{header_part}.{payload_part}"),
        signature,
        payload,
        kid: kid.to_owned(),
    })
}

/// The email and subject of an assertion whose signature already verified. The
/// refusal order is v2's (`cf-access.ts:227-254`) and each step is a distinct
/// reason. A claim of the WRONG TYPE refuses like an absent one.
fn validate_claims(
    payload: &[u8],
    domain: &str,
    audience: &str,
    now_ms: i64,
) -> Result<(String, String), AccessRejection> {
    let claims: Value = serde_json::from_slice(payload).map_err(|_| AccessRejection::BadClaims)?;
    let text = |name: &str| claims.get(name).and_then(Value::as_str).unwrap_or_default();
    if text("iss") != issuer_of(domain) {
        return Err(AccessRejection::BadIssuer);
    }
    // `aud` must be an ARRAY holding this tag; v2 refuses a bare string.
    let audience_matches = claims
        .get("aud")
        .and_then(Value::as_array)
        .is_some_and(|list| list.iter().any(|entry| entry.as_str() == Some(audience)));
    if !audience_matches {
        return Err(AccessRejection::BadAudience);
    }
    // `claim_instant_ms` returns MILLISECONDS because that is the unit a JWT
    // instant is compared against; the clock is converted to that unit ONCE
    // here, so no comparison mixes an integer millisecond instant with a float
    // one and the skew is a single named value rather than three casts.
    let now = now_ms as f64;
    let skew = ACCESS_CLOCK_SKEW_MS as f64;
    let instant = |name: &str| claim_instant_ms(claims.get(name).and_then(Value::as_f64));
    if instant("exp")? <= now - skew {
        return Err(AccessRejection::Expired);
    }
    if instant("iat")? > now + skew {
        return Err(AccessRejection::BadClaims);
    }
    if let Some(not_before) = claims.get("nbf").and_then(Value::as_f64)
        && claim_instant_ms(Some(not_before))? > now + skew
    {
        return Err(AccessRejection::BadClaims);
    }
    let (email, subject) = (text("email"), text("sub"));
    if !is_reportable_email(email) || !is_reportable_identity_text(subject) {
        return Err(AccessRejection::BadClaims);
    }
    Ok((email.to_owned(), subject.to_owned()))
}

/// The issuer a claim's `iss` must equal for this team domain.
fn issuer_of(domain: &str) -> String {
    format!("https://{domain}")
}

/// A claim instant in milliseconds. The conversion must be finite as well as the
/// claim: an infinite `exp` satisfies "not yet expired" against any clock.
fn claim_instant_ms(seconds: Option<f64>) -> Result<f64, AccessRejection> {
    let finite = seconds
        .filter(|value| value.is_finite())
        .ok_or(AccessRejection::BadClaims)?;
    let millis = finite * 1_000.0;
    millis
        .is_finite()
        .then_some(millis)
        .ok_or(AccessRejection::BadClaims)
}

/// Whether a string may be recorded as an identity, whatever it claims to be.
///
/// STRICTER THAN v2, ON PURPOSE. `cf-access.ts:247-253` bounds `email` only by
/// non-empty and 320 UTF-8 bytes, and that text reaches an audit row, a
/// `pair_requests` row and every browser's device list -- so a newline or a bidi
/// override in it is log forging.
#[must_use]
pub fn is_reportable_identity_text(value: &str) -> bool {
    if value.is_empty() || value.trim() != value {
        return false;
    }
    !value.chars().any(|character| {
        character.is_control()
            || matches!(
                character,
                '\u{200b}'..='\u{200f}'
                    | '\u{202a}'..='\u{202e}'
                    | '\u{2066}'..='\u{2069}'
                    | '\u{feff}'
            )
    })
}

/// An address a pairing request may record: [`is_reportable_identity_text`] plus one `@`,
#[must_use]
pub fn is_reportable_email(value: &str) -> bool {
    if !is_reportable_identity_text(value) || value.len() > MAX_ACCESS_EMAIL_UTF8_BYTES {
        return false;
    }
    let Some((local, domain)) = value.split_once('@') else {
        return false;
    };
    let whitespace_free = |part: &str| part.chars().all(|c| !c.is_whitespace());
    !local.is_empty()
        && !domain.is_empty()
        && !domain.contains('@')
        && whitespace_free(local)
        && whitespace_free(domain)
}
